modules/git/command.go

```diff
@@ -292,12 +292,13 @@ func (c *Command) RunStdBytes(opts *RunOpts) (stdout, stderr []byte, runErr RunS
 }
 
 // AllowLFSFiltersArgs return globalCommandArgs with lfs filter, it should only be used for tests
+// It also re-enables git-credential(1), which is used to test git-annex's HTTP support
 func AllowLFSFiltersArgs() []string {
 	// Now here we should explicitly allow lfs filters to run
 	filteredLFSGlobalArgs := make([]string, len(globalCommandArgs))
 	j := 0
 	for _, arg := range globalCommandArgs {
-		if strings.Contains(arg, "lfs") {
+		if strings.Contains(arg, "lfs") || strings.Contains(arg, "credential") {
```

Comment:
What a strange function! I really don't understand how it's supposed to work. At least the description says it's only used during tests? I sure hope it's not used in production.

* It uses `strings.Contains` to test for a _substring_, not for string equality, so for example a filename which just contains the three letters `lfs` (or now, the ten letters `credential`) will pass this test.
* The loop doesn't just filter out the arguments which pass the test, but `j--` means that the previous argument gets discarded as well.
* In fact, if several arguments pass the test, the index `j` could very well become negative and be out of bounds.

Reply:
I know! It's nuts!

This is used to counteract this chunk:

https://github.com/neuropoly/gitea/blob/65d8c81e537da1bc6919e91c04ae6a12e6045260/modules/git/git.go#L167-L203

We could sidestep it by using `NewCommandNoGlobals()` directly:

https://github.com/neuropoly/gitea/blob/65d8c81e537da1bc6919e91c04ae6a12e6045260/modules/git/command.go#L65-L76

however the stuff I'm reusing in the tests like `doGitClone()` calls `AllowLFSFiltersArgs()` directly:

https://github.com/neuropoly/gitea/blob/65d8c81e537da1bc6919e91c04ae6a12e6045260/tests/integration/git_helper_for_declarative_test.go#L128-L130

which of course starts from `globalCommandArgs`:

https://github.com/neuropoly/gitea/blob/65d8c81e537da1bc6919e91c04ae6a12e6045260/modules/git/command.go#L298

I think it's safe-ish, it's only ever applied to stuff in `globalCommandArgs`, which is only ever set (:crossed_fingers:) in `InitFull()`.

Reply:
But yeah why they didn't just use `append()` is beyond me.

Reply:
That makes sense! Thanks for the snippet of `git.go`, it was the missing context for me. I'm still glad this code is only used for testing :wink:

Reply:
Oh I see why. Because 

```
globalCommandArgs = []string{"-c", "protocol.version=2", "-c", "credential.helper=", "-c", "filter.lfs.required=", "-c", "filter.lfs.smudge=", "-c", "filter.lfs.clean="}
```

i.e. arguments come in pairs, and you need to examine the future argument before deciding to append `-c` or not. So they did that with this weird `j` thing. `j` can't go negative so long as there's always a `-c`.

Which still seems awkward but whatever.